modules/@apostrophecms/log/index.js

```diff
@@ -0,0 +1,403 @@
+      },
+
+      // Enrich the `data` argument with additional information from the request.
+      processRequestData(req, data) {
+        if (!req) {
+          return data;
+        }
+
+        // Keep the proper property order for better readability.
+        const message = {};
+        if (self.options.messageAs) {
+          message[self.options.messageAs] = data[self.options.messageAs];
+        }
+        const {
+          module: _module, type, severity, ...rest
+        } = data;
+
+        return {
+          ...message,
+          module: _module,
+          type,
+          severity,
+          // https://expressjs.com/en/api.html#req.originalUrl
+          url: req.originalUrl,
+          path: req.path,
+          method: req.method,
+          ip: req.ip,
+          query: req.query,
+          requestId: req.requestId || self.apos.util.generateId(),
+          ...rest
+        };
+      },
+
+      // Assess the module filter configuration and determine if the log
+      // should be kept or rejected.
+      //
+      // `moduleSelf` and `args` arguments should be the same as
+      // passed to `processLogArgs(...)`.
+      //
+      // The module and global configs are merged.
+      // The logic is as follows:
+      // - if severity is matched, keep the log (no matter the event type)
+      // - if eventType is matched, keep the log (no matter the severity)
+      shouldKeepEntry(moduleSelf, ...args) {
+        // Detect severity and eventType from the arguments.
+        let severity;
+        let req;
+        let eventType;
+        if (args[1] && typeof args[1].t === 'function') {
+          // eslint-disable-next-line no-unused-vars
+          [ severity, req, eventType ] = args;
+        } else {
+          [ severity, eventType ] = args;
+        }
+        const module = moduleSelf.__meta?.name ?? '__unknown__';
+
+        // Consolidate and match severity and event type.
+        const severityArr = [
+          ...new Set([
+            ...self.filters['*'].severity,
+            ...(self.filters[module]?.severity || [])
+          ])
+        ];
+        const eventsArr = [
+          ...new Set([
+            ...(self.filters['*'].events || []),
+            ...(self.filters[module]?.events || [])
+          ])
+        ];
+        const severityMatch = severityArr.includes(severity);
+        const eventsMatch = eventsArr.length > 0
+          ? eventsArr.includes(eventType) || eventsArr.includes('*')
+          : severityMatch;
+
+        return severityMatch || eventsMatch;
+      },
+
+      // Stringify object arguments. If the environment is not `production`,
+      // pretty print the objects and add a new line at the end of string arguments.
+      // This method is meant to be used from the methods of a custom `logger`.
+      // See the default logger implementation in `util/lib/logger.js` for an example.
+      formatLogByEnv(args) {
+        const formatObj = process.env.NODE_ENV === 'production'
```

Comment:
Don't make new functions on every call; check NODE_ENV just once at startup and assign one of them to use going forward for the lifetime of the process.

Reply:
So true :)